### PR TITLE
[v2] Breaking: Change async validate to resolve errors

### PR DESCRIPTION
**Pull request by jaredpalmer**
- async `<Formik validate />` and `<Field validate>` should now resolve errors object or string (instead of throwing it). IIRC I kept redux-form's API when adding this initially, but this is dumb because all errors have been swallowed.
- `validationSchema` will now throw non-yup validation errors instead of swallowing them.

Both of these are breaking changes, but they are relatively small.

### CODE REVIEW

**Comment by jaredpalmer** [comment]
I wasn't sure about rejecting vs. throwing some of these. anybody have an opinion? @Andreyco

File: src/Formik.tsx [hunk 1]
```diff
@@ -390,7 +413,7 @@ function useFormikInternal<Values = object>({
           // Only flip isValidating if the function is async.
           dispatch({ type: 'SET_ISVALIDATING', payload: true });
           return maybePromise
-            .then((x: any) => x, (e: any) => e)
+            .then((x: any) => x)
```
Comment (Andreyco) [inline src/Formik.tsx]: 👏

**Comment by Andreyco** [comment]
I like current approach, when only resolved value means validation run successfully (regardless its output). I love validation handler does not swallow exceptions anymore.

PS1: While breaking this handler for v2, did you consider `undefined` or `null` for "no errors" instead of empty `{}`?

PS2: How about dropping async validation as proposed before?

File: src/Formik.tsx [hunk 2]
```diff
@@ -162,17 +162,24 @@ function useFormikInternal<Values = object>({
...
+        if (!maybePromisedErrors) {
           resolve(emptyErrors);
         } else if (isPromise(maybePromisedErrors)) {
           (maybePromisedErrors as Promise<any>).then(
-            () => {
-              resolve(emptyErrors);
-            },
             errors => {
-              resolve(errors);
+              resolve(errors || {});
```
[the remark this answers is not in the document]
Comment (jaredpalmer) [reply src/Formik.tsx]: Good catch.

**Comment by jaredpalmer** [comment]
PS1 is a huge breaking change. We've been showing folks that they can just do `errors.firstName && touched.firstName` since v0. They would have to add a check everywhere (unless they were using `getIn`. It also could cause confusion, since there would be multiple ways to express the idea of "no errors": `{}` or `undefined`.

PS2: I'm experimenting with that now

PS3: I'm not sure what yup validator binding is? Is that the context argument?

**Comment by jaredpalmer** [comment]
@Andreyco any idea how we make the tests not actually throw?

**Comment by Andreyco** [comment]
> PS1 is a huge breaking change. We've been showing folks that they can just do `errors.firstName && touched.firstName` since v0. They would have to add a check everywhere (unless they were using `getIn`. It also could cause confusion, since there would be multiple ways to express the idea of "no errors": `{}` or `undefined`.
> 

What I meant is to return void from validation handler. I am perfectly fine with having `formik.errros = {}`. Thinking about this now, and I am perfectly okay with returning `{}` from validation handler. Please, disregard.


> PS2: I'm experimenting with that now

👏 

> PS3: I'm not sure what yup validator binding is? Is that the context argument?

I meant `validationSchema` prop, which accepts Yup's validation schema/validator. There are many issues and PRs that aim to extend Yups integration. While I am not against this integration, I am against extending Formik's API, especially if this can be solved completely in user land.

> @Andreyco any idea how we make the tests not actually throw?

Will have a look

**Comment by jaredpalmer** [comment]
I agree. I am against any PR expanding Yup integration

**Comment by jaredpalmer** [comment]
Hrmmm. What should happen when you return void? Keep errors? or set them to `{}`

**Pull request merged**